Features/fast overview

ppillot opened the pull request:
This branch changes the inner workings of the overview REST API endpoint by using DB queries to compute statistics (previously it was doing a single query with multiple joins and computing statistics by parsing the tables of results).

**Intentional change**: the statistics about age are returned already binned. This can be reverted if deemed needed.

**How to test:**
- ensure that the previous response and this one are identical. This can be done by loading the overview page in a browser and inspecting the network

**Known issue:**
~~In some cases the `age` property can be a range. This is not handled by the current code.~~

Commit:
convert overview counters to DB queries

Commit:
implement binned ages

Commit:
add test for age computation in overview

Commit:
compute age stats when age_numeric is missing

ppillot commented:
Age defined as ranges are handled in the latest iteration

gsfk commented:
Nice! This is _much_ faster, where the old version was already too slow. 

This version bins ages by decade, where the previous version binned by year, which required an extra step of binning by decade in the front end. I think the new approach is preferable. 

Still a couple bugs to work out, though, since local tests show a few quirks: 
- all diseases have a count of one
- all instrument platforms have a count of one
- sampled tissue counts disagree with counts given by the current katsu verison (and for the handful that I checked the current version was correct)

gsfk reviewed (commented):
Nice! This is _much_ faster, where the old version was already too slow. 

This version bins ages by decade, where the previous version binned by year, which required an extra step of binning by decade in the front end. I think the new approach is preferable. 

Still a couple bugs to work out, though, since local tests show a few quirks: 
- all diseases have a count of one
- all instrument platforms have a count of one
- sampled tissue counts disagree with counts given by the current katsu verison (and for the handful that I checked the current version was correct)

Commit:
missing joins for stats

ppillot commented:
@gsfk Thanks for your tests! These highlight issues with the synthetic dataset as well: there is only one disease referenced for only one patient...
The disease and the instruments related stats have been fixed now.
Regarding the Biosamples, let's discuss it: I currently don't understand where is the issue... there is something I probably miss in my comprehension of the relationship between phenopackets and biosamples.

Commit:
add more experiments to tests

Commit:
fix: duplicated labels on distinct fields counted once

By aggregating results on the label directly the counts
from the database are made exact.
The code for dictionnary generation can be simplified

Commit:
fix test

It was breaking on another test where the expected experiment id could
not be found due to a change in nums.

Commit:
fix diseases count

Was also counting diseases defined in the DB but not rereferenced
in any phenopackets.

SanjeevLakhwani commented on chord_metadata_service/restapi/api_views.py:
```diff
@@ -44,173 +47,87 @@ def overview(_request):
-                "experiment_type": dict(experiments["experiment_type"]),
-                "molecule": dict(experiments["molecule"]),
-                "library_strategy": dict(experiments["library_strategy"]),
-                "library_source": dict(experiments["library_source"]),
-                "library_selection": dict(experiments["library_selection"]),
-                "library_layout": dict(experiments["library_layout"]),
-                "extraction_protocol": dict(experiments["extraction_protocol"]),
+                "count": experiments_count,
+                "study_type": stats_for_field(experiments_models.Experiment, "study_type"),
+                "experiment_type": stats_for_field(experiments_models.Experiment, "experiment_type"),
+                "molecule": stats_for_field(experiments_models.Experiment, "molecule"),
+                "library_strategy": stats_for_field(experiments_models.Experiment, "library_strategy"),
+                "library_source": stats_for_field(experiments_models.Experiment, "library_source"),
+                "library_selection": stats_for_field(experiments_models.Experiment, "library_selection"),
+                "library_layout": stats_for_field(experiments_models.Experiment, "library_layout"),
+                "extraction_protocol": stats_for_field(experiments_models.Experiment, "extraction_protocol"),
             },
             "experiment_results": {
-                "count": len(experiment_results_set),
-                "file_format": dict(experiment_results["file_format"]),
-                "data_output_type": dict(experiment_results["data_output_type"]),
-                "usage": dict(experiment_results["usage"])
+                "count": experiment_results_count,
+                "file_format": stats_for_field(experiments_models.ExperimentResult, "file_format"),
+                "data_output_type": stats_for_field(experiments_models.ExperimentResult, "data_output_type"),
+                "usage": stats_for_field(experiments_models.ExperimentResult, "usage")
             },
             "instruments": {
-                "count": len(instrument_set),
-                "platform": dict(instruments["platform"]),
-                "model": dict(instruments["model"])
+                "count": instruments_count,
+                "platform": stats_for_field(experiments_models.Experiment, "instrument__platform"),
+                "model": stats_for_field(experiments_models.Experiment, "instrument__model")
             },
         }
-    })
+    }
+
+    return Response(r)
```
Just wondering, the motivation for changing it to a variable instead of a direct response

ppillot replied:
Easier to debug by placing breakpoints. Can be simplified.

SanjeevLakhwani commented on chord_metadata_service/restapi/api_views.py:
```diff
@@ -452,3 +369,48 @@ def sort_numeric_values_into_bins(values: dict, bin_size: int = 10, threshold: i
+    # Count() aggregates the results by performing a GROUP BY on the field
+    query_set = model.objects.all().values(field).annotate(total=Count(field))
+    stats = dict()
+    for item in query_set:
+        key = item[field]
+        if key is None:
+            continue
+
+        key = key.strip()
+        if key == "":
+            continue
+
+        stats[key] = item["total"]
+    return stats
+
+
+def get_field_bins(model, field, bin_size):
+    # creates a new field "binned" by substracting the modulo by bin size to
+    # the value which requires binning (e.g. 28 => 28 - 28 % 10 = 20)
+    # cast to integer to avoid numbers such as 60.00 if that was a decimal,
+    # and aggregate over this value.
+    query_set = model.objects.all().annotate(
+        binned=Cast(
+            F(field) - Func(F(field), bin_size, function="MOD"),
+            IntegerField()
+        )
+    ).values('binned').annotate(total=Count('binned'))
+    stats = {item['binned']: item['total'] for item in query_set}
+    return stats
+
+
+def compute_binned_ages(bin_size):
+    a = pheno_models.Individual.objects.filter(age_numeric__isnull=True).values('age')
+    binned_ages = []
+    for r in a.iterator():  # reduce memory footprint (no caching)
+        if r["age"] is None:
+            continue
+        age = parse_individual_age(r["age"])
+        binned_ages.append(age - age % OVERVIEW_AGE_BIN_SIZE)
+    return binned_ages
```
Might be better to have them nested within the response function, or have a utility function set file for it, as then the file would have a better logical arrangement that each function as an api response function and not an util function

SanjeevLakhwani reviewed (commented):
Looks great, I do not have anything to add on but just 2 discussions. Great work! and Love the changes. I was also wondering if using numpy would be an option for the project as that would have faster mathematical calculation speeds for functions such as compute_binned_ages and possibly at other places too.
